js/src/Ice/ConnectionI.js

```diff
@@ -1077,21 +1086,54 @@ export class ConnectionI {
         }
     }
 
-    sendHeartbeatNow() {
-        Debug.assert(this._state === StateActive);
+    idleCheck(idleTimeout, rescheduleTimer) {
+        if (this._state == StateActive || this._state == StateHolding) {
+            if (this._transceiver.isWaitingToBeRead()) {
+                // Bytes are available for reading but we the event loop hans't processed them yet. We don't want to abort
```

Comment:
It's not clear to me how flow control / back pressure works in Ice for JS.
When do we stop reading more data/frames from the connection?

Reply:
For JavaScript browser WebSockets, there aren't any APIs to stop reading. For sending, we use the bufferedAmount property.

We process the data events as they come. Because JavaScript is single-threaded, we process one event at a time. The rest are queued in the event loop, and it is up to the browser implementation to apply back pressure.

In NodeJS, we might want to call pause on the Socket, but currently, we do not.

[See](https://github.com/zeroc-ice/ice/blob/5eed18656148bafdbb25bdb044ffa983db14c93c/js/src/Ice/TcpTransceiver.js#L253-L256)

Reply:
There is a proposal to fix the WebSocket case https://developer.chrome.com/docs/capabilities/web-apis/websocketstream

Reply:
Since we never stop reading, I suspect we don't need `isWaitingToBeRead`.

Worded differently: you won't be able to write a test where isWaitingToBeRead() returns false. Assuming you could write `testIdleCheckDoesNotAbortConnectionWhenThreadPoolIsExhausted` all in JS, the test will succeed because you keep reading / dispatching the heartbeat frames.

Reply:
I see. In JavaScript, once the dispatch is awaited, control returns to the event loop, which will continue dispatching heartbeat frames.

If the dispatch remains busy and does not relinquish control, we would not be able to process the Idle timer event either.

Reply:
Removed isWaitingToBeRead